### PR TITLE
Mmr persist state

### PR DESCRIPTION
**Pull request by acatangiu**
Persisting gadget state across runs.

Fixes https://github.com/paritytech/substrate/issues/12780

### CODE REVIEW

File: client/merkle-mountain-range/src/offchain_mmr.rs [hunk 1]
```diff
@@ -174,16 +200,58 @@ where
...
+					Some(header) => header,
+					_ => break,
+				};
+				if header.number <= self.best_canonicalized {
+					break
+				}
+				to_canon.push_front(header.hash);
+			}
+			// Canonicalize all blocks leading up to current finality notification.
+			for hash in to_canon.drain(..) {
```
Comment (svyatonik) [inline client/merkle-mountain-range/src/offchain_mmr.rs:232]: Sorry for the dumb q, but is the canonicalization order that important here? If it isn't important then we can do it directly in previous loop without this temp`to_canon` vec. But maybe it breaks something in MMR structure and then this is required - just wanted to check

File: client/merkle-mountain-range/src/aux_schema.rs [hunk 2]
```diff
@@ -0,0 +1,228 @@
+// This file is part of Substrate.
```
Comment (svyatonik) [inline client/merkle-mountain-range/src/aux_schema.rs:1]: I've had deja vu looking at this file (I mean #12712 ) :)

**Review by svyatonik — APPROVED** [review]
LGTM

File: client/merkle-mountain-range/src/offchain_mmr.rs (hunk 1 above)
Comment (acatangiu) [reply client/merkle-mountain-range/src/offchain_mmr.rs:232]: we're saving `best_canonicalized` in the canonicalization function (with no checks, assuming block numbers always increasing), and also logging a warning when we see "skipped" blocks - that check is currently very simple: `if current != best.saturating_add(1) { warn!() }`;

the logic is kept simple in the canonicalization function with the assumption that blocks are canonicalized in order

**Review by serban300 — APPROVED** [review]

File: client/merkle-mountain-range/src/aux_schema.rs [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+use sp_runtime::traits::{Block, NumberFor};
+
+const VERSION_KEY: &[u8] = b"mmr_auxschema_version";
+const GADGET_STATE: &[u8] = b"mmr_gadget_state";
+
+const CURRENT_VERSION: u32 = 1;
+pub(crate) type PersistedState<B> = NumberFor<B>;
+
+pub(crate) fn write_current_version<B: AuxStore>(backend: &B) -> ClientResult<()> {
+	info!(target: LOG_TARGET, "🥩 write aux schema version {:?}", CURRENT_VERSION);
```
[the remark this answers is not in the document]
Comment (acatangiu) [reply client/merkle-mountain-range/src/aux_schema.rs:35]: right 😆

**Pull request merged**